src/actions/threads.ts

```diff
@@ -0,0 +1,111 @@
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        let userThreadList: undefined | UserThreadList;
+
+        try {
+            userThreadList = await Client4.getUserThreads(userId, teamId, {page, pageSize: perPage, extended: true});
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+            return {error};
+        }
+
+        dispatch({
+            type: PostTypes.RECEIVED_POSTS,
+            data: {posts: userThreadList.threads.map(({post}) => post)},
+        });
+
+        dispatch({
+            type: ThreadTypes.RECEIVED_THREADS,
+            data: {
+                ...userThreadList,
+                threads: userThreadList.threads.map((thread) => ({...thread, is_following: true})),
+                team_id: teamId,
+            },
+        });
+
+        return {data: userThreadList};
+    };
+}
+
+export function markAllThreadsInTeamRead(userId: string, teamId: string) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        try {
+            await Client4.updateThreadsReadForUser(userId, teamId);
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+            return {error};
+        }
+
+        dispatch({
+            type: ThreadTypes.ALL_TEAM_THREADS_READ,
+            data: {
+                team_id: teamId,
+            },
+        });
+
+        return {};
+    };
+}
+
+export function updateThreadRead(userId: string, teamId: string, threadId: string, timestamp: number) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        try {
+            await Client4.updateThreadReadForUser(userId, teamId, threadId, timestamp);
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+            return {error};
+        }
+
+        dispatch({
+            type: ThreadTypes.READ_CHANGED_THREAD,
+            data: {
+                id: threadId,
+                team_id: teamId,
+                timestamp,
+            },
+        });
+
+        return {};
+    };
+}
+
+export function setThreadFollow(userId: string, teamId: string, threadId: string, newState: boolean) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        try {
+            await Client4.updateThreadFollowForUser(userId, teamId, threadId, newState);
+        } catch (error) {
+            forceLogoutIfNecessary(error, dispatch, getState);
+            dispatch(logError(error));
+            return {error};
+        }
+
+        dispatch({
+            type: ThreadTypes.FOLLOW_CHANGED_THREAD,
+            data: {
+                id: threadId,
+                team_id: teamId,
+                following: newState,
+            },
+        });
```

Comment:
Should we dispatch this lazily and then again if need be after success, to prevent this on slow connections?
![CleanShot 2021-01-11 at 08 51 26](https://user-images.githubusercontent.com/11724372/104197110-39aca300-53ea-11eb-86bd-b95f6563548f.gif)